source/secure_tunneling.c

```diff
@@ -473,198 +396,1395 @@ bool secure_tunneling_send_data_call(struct aws_websocket *websocket, struct aws
+        secure_tunnel->next_service_task_run_time > 0) {
+        aws_event_loop_cancel_task(secure_tunnel->loop, &secure_tunnel->service_task);
+        secure_tunnel->next_service_task_run_time = 0;
+
+        AWS_LOGF_TRACE(
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: cancelling previously scheduled service task",
+            (void *)secure_tunnel);
+    }
+
+    if (next_service_time > 0 && (next_service_time < secure_tunnel->next_service_task_run_time ||
+                                  secure_tunnel->next_service_task_run_time == 0)) {
+        aws_event_loop_schedule_task_future(secure_tunnel->loop, &secure_tunnel->service_task, next_service_time);
+
+        AWS_LOGF_TRACE(
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: scheduled service task for time %" PRIu64,
+            (void *)secure_tunnel,
+            next_service_time);
+    }
+
+    secure_tunnel->next_service_task_run_time = next_service_time;
+}
+
+/*********************************************************************************************************************
+ * Update Loop
+ ********************************************************************************************************************/
+
+static int s_aws_secure_tunnel_queue_ping(struct aws_secure_tunnel *secure_tunnel) {
+    s_reset_ping(secure_tunnel);
+
+    AWS_LOGF_DEBUG(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "id=%p: queuing PING", (void *)secure_tunnel);
+
+    struct aws_secure_tunnel_operation_pingreq *pingreq_op =
+        aws_secure_tunnel_operation_pingreq_new(secure_tunnel->allocator);
+    s_enqueue_operation_front(secure_tunnel, &pingreq_op->base);
+
+    return AWS_OP_SUCCESS;
+}
+
+static bool s_service_state_stopped(struct aws_secure_tunnel *secure_tunnel) {
+    enum aws_secure_tunnel_state desired_state = secure_tunnel->desired_state;
+    if (desired_state == AWS_STS_CONNECTED) {
+        s_change_current_state(secure_tunnel, AWS_STS_CONNECTING);
+    } else if (desired_state == AWS_STS_TERMINATED) {
+        s_change_current_state(secure_tunnel, AWS_STS_TERMINATED);
+        return true;
+    }
+    return false;
+}
+
+static void s_service_state_connecting(struct aws_secure_tunnel *secure_tunnel, uint64_t now) {
+    (void)secure_tunnel;
+    (void)now;
+}
+
+static void s_service_state_connected(struct aws_secure_tunnel *secure_tunnel, uint64_t now) {
+    enum aws_secure_tunnel_state desired_state = secure_tunnel->desired_state;
+    if (desired_state != AWS_STS_CONNECTED) {
+        AWS_LOGF_INFO(
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: channel shutdown due to user Stop request",
+            (void *)secure_tunnel);
+        s_secure_tunnel_shutdown_websocket(secure_tunnel, AWS_ERROR_IOTDEVICE_SECURE_TUNNELING_USER_REQUESTED_STOP);
+        return;
+    }
+
+    if (now >= secure_tunnel->next_ping_time) {
+        if (s_aws_secure_tunnel_queue_ping(secure_tunnel)) {
```

Comment:
Minor: Unless I am mistaken, this function (`s_aws_secure_tunnel_queue_ping`) cannot fail? It is supposed to be able to fail?

Reply:
There's no current way for it to fail. But maybe someday...